File: python/taflow/indicators/new_highs_new_lows.py
"""Python adapter for native New Highs minus New Lows breadth."""

from typing import Any
import numpy as np
from .._native import NewHighsNewLows as _Native
from .._series import as_float64_series


class NewHighsNewLows:
    """Return ``new_highs - new_lows`` for each cross-sectional tick.

    Inputs are pre-aggregated constituent counts. The definition maps directly
    to Wickra ``NewHighsNewLows`` and emits from the first tick.

    Args:
        new_highs: Number of constituents making a new high.
        new_lows: Number making a new low at each aligned tick.

    Raises:
        ValueError: If the two histories have different lengths.
    """

    def __init__(self, new_highs: Any, new_lows: Any) -> None:
        """Initialize native state and process aligned count histories."""
        self._state = _Native()
        self.extend(new_highs, new_lows)

    def append(self, new_highs: float, new_lows: float) -> "NewHighsNewLows":
        """Append one aggregate-extremes tick and return this adapter."""
        self._state.append(float(new_highs), float(new_lows))
        return self

    def extend(self, new_highs: Any, new_lows: Any) -> "NewHighsNewLows":
        """Append aligned count histories after validating their lengths."""
        arrays = as_float64_series(new_highs), as_float64_series(new_lows)
        if len(arrays[0]) != len(arrays[1]):
            raise ValueError("new-high and new-low counts must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest net count, or ``None`` before input."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return the aligned net new-extremes history."""
        return self._state.compute()

    def reset(self) -> "NewHighsNewLows":
        """Reset native state and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-tick count delegated to native state."""
        return len(self._state)

File: crates/taflow-core/src/indicators/new_highs_new_lows.rs
use crate::error::TaResult;
/// Difference between cross-sectional new-high and new-low counts.
#[derive(Debug, Clone, Default)]
pub struct NewHighsNewLows {
    count: usize,
    value: Option<f64>,
}
impl NewHighsNewLows {
    /// Create an empty net-extremes state.
    pub fn new() -> TaResult<Self> {
        Ok(Self::default())
    }
    /// Append aggregate new-high and new-low counts for one market tick.
    pub fn append(&mut self, new_highs: f64, new_lows: f64) -> Option<f64> {
        self.count += 1;
        self.value = Some(new_highs - new_lows);
        self.value
    }
    /// Return the latest net new-extremes count.
    pub fn value(&self) -> Option<f64> {
        self.value
    }
    /// Return the number of processed market ticks.
    pub fn len(&self) -> usize {
        self.count
    }
    /// Return whether no market ticks have been processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }
    /// Restore fresh-state behaviour.
    pub fn reset(&mut self) {
        self.count = 0;
        self.value = None;
    }
}
